### PR TITLE
add autoComplete search

### PR DESCRIPTION
**Pull request by ddis**

### CODE REVIEW

File: src/Entities/AutoCompleteEntity.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: User
```
Comment (comaw) [inline src/Entities/AutoCompleteEntity.php]: User - Who are you?

File: src/Entities/AutoCompleteEntity.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * @property int    $id
+ * @property string $title
+ * @property string $urlCode
+ */
+abstract class AutoCompleteEntity extends Entity
+{
+    public $id;
+    public $title;
+    public $urlCode;
+}
```
Comment (comaw) [inline src/Entities/AutoCompleteEntity.php]: No newline at end file

File: src/Lib/Filters/ElasticSearch/AutoCompleteFilter.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * @param bool $isHighlight
+     * @return array
+     */
+    private function combineParams($isHighlight = false)
+    {
+        $result = [];
+
+        foreach ($this->values as $key => $value) {
+            foreach ($this->langCodes as $langCode) {
```
Comment (comaw) [inline src/Lib/Filters/ElasticSearch/AutoCompleteFilter.php]: to my opinion - you need a function for nested foreach

File: src/Lib/Filters/Interfaces/AutoCompleteInterface.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * Interface AutoCompleteInterface
+ * @package SphereMall\MS\Lib\Filters\Interfaces
+ */
+interface AutoCompleteInterface
+{
+    /**
+     * @return mixed
+     */
+    public function getHighlight();
+}
```
Comment (comaw) [inline src/Lib/Filters/Interfaces/AutoCompleteInterface.php:21]: No newline at end file

File: src/Lib/Makers/ElasticSearchAutoCompleteMaker.php [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            $mapperClass = $this->getMapperClass($element['_type']);
+            if (is_null($mapperClass)) {
+                throw new EntityNotFoundException("Entity mapper class for {$element['_type']} was not found");
+            }
+
+            if (!$element['_source']) {
+                continue;
+            }
+
+            foreach ($this->langCodes as $language) {
```
Comment (comaw) [inline src/Lib/Makers/ElasticSearchAutoCompleteMaker.php:56]: to my opinion - you need a function for nested foreach

File: src/Lib/Filters/ElasticSearch/AutoCompleteFilter.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            'post_tags' => $this->highLightTagPost,
+            'fields'    => $this->combineParams(true),
+        ];
+    }
+
+    /**
+     * @param bool $isHighlight
+     * @return array
+     */
+    private function combineParams($isHighlight = false)
```
Comment (comaw) [inline src/Lib/Filters/ElasticSearch/AutoCompleteFilter.php]: private - no good idea to free script

File: src/Resources/ElasticSearch/ElasticSearchResource.php [hunk 7]
```diff
@@ -186,5 +200,21 @@ public function delete($id)
...
+    #endregion
+
+    #region [privates methods]
+    /**
+     * @param $params
+     * @return \GuzzleHttp\Promise\PromiseInterface|ElasticSearchResponse|\SphereMall\MS\Lib\Http\Response
+     * @throws \Exception
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    private function getResponse($params)
```
Comment (comaw) [inline src/Resources/ElasticSearch/ElasticSearchResource.php:213]: private - no good idea to free script

Comment (comaw) [reply src/Resources/ElasticSearch/ElasticSearchResource.php:213]: And maybe 'GET' - need var?

File: tests/_source/autoComplete.json [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    "total"    : 4,
+    "max_score": 0,
+    "hits"     : [
+      {
+        "_index"   : "sm-products",
+        "_type"    : "products",
+        "_id"      : "6354",
+        "_score"   : 0,
+        "_source"  : {
+          .
```
Comment (comaw) [inline tests/_source/autoComplete.json]: not valide json file?

File: src/Lib/Makers/ElasticSearchAutoCompleteMaker.php [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+// ToDo: extends Maker or ObjectMaker with overriding
+
+/**
+ * Class ElasticSearchMaker
+ * @package SphereMall\MS\Lib\Makers
+ *
+ * @property bool $asCollection
+ */
+class ElasticSearchAutoCompleteMaker
```
Comment (SergeyKondratovec) [inline src/Lib/Makers/ElasticSearchAutoCompleteMaker.php]: Class is not extended

File: src/Lib/Makers/ElasticSearchAutoCompleteMaker.php [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        return $result;
+    }
+
+    /**
+     * @param $type
+     * @return null|string
+     */
+    protected function getMapperClass($type)
```
Comment (SergeyKondratovec) [inline src/Lib/Makers/ElasticSearchAutoCompleteMaker.php]: Looks like we have this code

File: src/Lib/Makers/ElasticSearchAutoCompleteMaker.php [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    public function __construct($langCodes = [])
+    {
+        $this->langCodes = $langCodes;
+    }
+    /**
+     * @param ElasticSearchResponse $response
+     * @return mixed|null
+     * @throws EntityNotFoundException
+     */
+    public function makeSingle(ElasticSearchResponse $response)
```
Comment (SergeyKondratovec) [inline src/Lib/Makers/ElasticSearchAutoCompleteMaker.php]: Should be Basic Response class

File: src/Entities/Page.php [hunk 12]
```diff
@@ -29,7 +29,7 @@
  * @property int            $noindex
  * @property FunctionalName $functionalName
  */
-class Page extends Entity
+class Page extends AutoCompleteEntity
 {
     #region [Properties]
     public $id;
```
Comment (SergeyKondratovec) [inline src/Entities/Page.php]: You did extend it

**Review by ghost — APPROVED** [review]

**Review by comaw — APPROVED** [review]

**Review by SergeyKondratovec — APPROVED** [review]

**Pull request merged**